### PR TITLE
Mk 243262 migration for new location fixture

### PR DESCRIPTION
**Pull request by mkangia**
[Spec](https://docs.google.com/document/d/1KjGPZL4oHyFUeAajnyNYII5Un276_ofc8WiTbWkoaYM/edit#)

🏛 

command to run 
`./manage.py migrate_new_location_fixture`
for dry run use `./manage.py migrate_new_location_fixture --check`

### CODE REVIEW

**Comment by mkangia** [comment]
@esoergel would add more checks for finding out applications using hierarchical fixture. But everything else should be in place. Updated Spec as well.

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (proteusvacuum) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: I think here you want to do `HIERARHICAL_LOCATION_FIXTURE.set(domain, enabled=True, namespace=NAMESPACE_DOMAIN)`, because that also updates the toggle cache

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (proteusvacuum) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: does this also find references in form xml?

File: corehq/tabs/tabclasses.py
Comment (mkangia) [inline corehq/tabs/tabclasses.py]: oops! that was simple.

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (mkangia) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: Not yet Farid. Waiting for some feedback for more areas to search. Included application doc only for now.

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (mkangia) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: need to correct that.. i intended to use `toggle.set` only.

Comment (czue) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: you can also just append the (namespaced) domain to the list of enabled items and save it at the end if we don't want to save this document once-per domain. you can just look at what `set_toggle` does behind the scenes

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (czue) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: I think the rendered complete set of source files might be the right thing to search in

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (czue) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: is there a chance we think we need to search app builds too? maybe just released ones? failure mode I'm thinking of is if someone has removed references in the current app version but still has users using earlier app versions.

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (czue) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: I think this should be `domain_with_application` not `search_string`

**Review by czue — CHANGES_REQUESTED** [review]
I think we need to be very careful with rolling this out. The failure mode here is that we unknowingly break a huge amount of people who are trying to use their applications. As-is if this PR is merged it will immediately do that.

I think we need to come up with a clear migration pathway that we can feel very confident is not going to break anyone, and I suspect we need to do that in multiple stages. I also think we will want a comprehensive test suite for the logic that checks for location references in applications so we can be sure to trust it.

It seems like it would be much easier to start by just making this the default for new domains, and converting everyone who is not using locations (no org hierarchy defined). Then focusing on figuring out what to do with the hard cases.

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (mkangia) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: Ideally would do that. I am not completely sure but seems like the cache expiry would get tricky there since that needs to be done for each domain.

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (mkangia) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: From the limited knowledge i have, in that case would this stand for all domains that migrate from old fixture format in an earlier version to flat fixture in new versions with users still using old version of app? May be because phone restore would still give new flat fixture format and app would be using old references?

Comment (czue) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: yeah exactly. so for example for TB we have eliminated hierarchical from our apps, but we have not turned off sync yet because one person is still using the old app

**Comment by esoergel** [comment]
It's a bit confusing to understand how the `FixtureConfiguration` and the toggles fit together, what is the intended interaction?

I just had an idea - what if you instead worked to drop both toggles and use only the `FixtureConfiguration` to determine what to sync?  This would mean creating a `FixtureConfiguration` for all domains currently using locations (we can be conservative and set `hierarchical=True` at first), then going forward, if a domain doesn't have a `FixtureConfiguration`, we sync only the flat fixture, otherwise we do what the `FixtureConfiguration` says.  That way it's stored clearly in one place, and there's a migration path for existing domains.  (This is just me spitballing, so feel free to ignore if it doesn't sound easier)

**Comment by czue** [comment]
@esoergel I like that idea a lot. The only issue is that for any app that is using both we do need to be explicit about which one we resolve in app manager. A couple thoughts:

1. we could add an option to the config and auto-set it safely during the migration
2. we could add as an option inside the app somewhere
3. we could try to detect it on build by inspecting your app contents (might be hard to implement and a performance hit - and we would have to figure out what to do if you referenced none or both of them in your app and likely need to fall back to 1 or 2 to resolve that...)

at a minimum I think making it explicit in the config object as opposed to mucking around with the feature flags will be much easier to reason with.

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (dimagimon) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: E203 whitespace before ':'

**Comment by mkangia** [comment]
@esoergel @czue have updated spec(Revised Changes V2) according to what was discussed last and also the PR accordingly. I have not tested it as of now. Once confirmed that the approach looks fine, will spend time testing it.
I have not looked in the idea above about dropping both toggles since that needs some more brainstorming.

**Comment by czue** [comment]
@mkangia I think it's worth making the chosen format in apps configurable as a part of this work and dropping both toggles. I don't think it should be hard - just use the same logic you are using in the toggles but insert the configuration object in place of it, and set it the same way you are setting the toggles.

**Comment by czue** [comment]
so to be clear I think the steps would be:

1. add a field "default format in apps" to `FixtureConfiguration` with allowed values ["flat", "hiearchical"] and default value "flat"
2. replace all toggle logic in apps to instead run off of this field
3. remove any toggle behavior from fixture generation code and rely only on the config object
4. update migration to set this field to "hieararchical" where you are currently setting the toggle.

@esoergel does that sound right?

**Comment by mkangia** [comment]
@czue if we drop both toggles, it would make the configuration option open for all users and would allow domains on flat fixture to move back to hierarchical configurations whereas we need the opposite to happen?

**Comment by czue** [comment]
good point - we should also hide the config UI if the app type is flat and only flat is turned on (which should also be the default for all new domains)

**Comment by mkangia** [comment]
@czue that sounds good. Just so i understand this correctly,
You are suggesting to keep the 2 fields `sync_flat_fixture` and `sync_hierarchical_fixture` as well along with `default format in apps`?

**Comment by czue** [comment]
> You are suggesting to keep the 2 fields `sync_flat_fixture` and `sync_hierarchical_fixture` as well along with `default format in apps`? 

yep exactly

**Comment by esoergel** [comment]
This all sounds great to me, I'm on the same page.

**Comment by esoergel** [comment]
There are some build failures, my guess is it's mostly just existing tests that need to be updated to reflect the new fixture.

File: corehq/ex-submodules/casexml/apps/phone/tests/test_ota_fixtures.py
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_ota_fixtures.py]: F401 'flag_enabled' imported but unused

**Comment by mkangia** [comment]
@esoergel I had a quick sync up with @czue and we concluded to stay on the approach that PR is following. I am having a tough time resolving the build failures. Needs a better knowledge of fixtures to fix them. Moreover, some build failures might highlight genuine concerns. Can you help with test cases? i have fixed  some of them but got stuck at some tricky ones now.

**Comment by mkangia** [comment]
Thank You @esoergel. :1st_place_medal:

**Review by esoergel — COMMENTED** [review]
I just did another review and it looks good to me.  Was there anything else you wanted to do before wrapping this up?  (The "don't merge" flag is still on).  @czue did you have any remaining followup?

**Comment by mkangia** [comment]
thanks @esoergel One last thing would be to ensure the places where `commtrack:location` exists in the code. I would assume that this notation should be used only for migration and other wise the new notation should be used everywhere.
As of now its mostly in test cases and just one place in code at https://github.com/dimagi/commcare-hq/blob/mk-243262-new-location-fixture/corehq/apps/app_manager/detail_screen.py#L614 which would be good check. Relying on your better understanding of fixtures. Let me know if any code/test-case i should specifically check still using the old notation.

**Comment by esoergel** [comment]
That looks like location xpath validation in the app builder, which I don't think I've ever seen in action.  @czue it looks like you wrote that back in 2014, any idea if that still works or where/how I'd be able to try it out?  [This](https://github.com/dimagi/commcare-hq/blob/3ccacef392e3c087ad4eab7c3a85179882d6538e/corehq/apps/app_manager/models.py#L2115) makes me think it's a case list thing, but I'm not sure how to get the option to appear.

**Comment by czue** [comment]
@esoergel @mkangia that is what powers [this functionality](https://confluence.dimagi.com/display/commtrack/Referencing+Locations+in+Applications). I think we can just get rid of that functionality if you're on the new fixture format. we probably need to support it for legacy apps though.

**Review by czue — APPROVED** [review]

**Comment by czue** [comment]
I haven't done a thorough review since last time, but I'm on board with the plan and signed off as long as you both are happy. Would be good to address / understand how this affects the behavior of that case list reference code before merging though.

**Comment by mkangia** [comment]
Thank You @esoergel @czue.
Will go ahead and try to do the same check for case list reference as well then.

@proteusvacuum Just checking if you did you come across this case list reference when working/moving to flat fixture locations.

**Comment by czue** [comment]
> @proteusvacuum Just checking if you did you come across this case list reference when working/moving to flat fixture locations.

there is no way it is currently working for flat fixture so i think we can conclude no one is using it.

**Comment by esoergel** [comment]
@mkangia I just tried this out and saw that it does indeed still exist and work.  The feature is pretty hidden though.  Go to a module on an app that uses cases (on a domain with locations).  Then add a new property, and type in something like `location:state:district/name` (where `state` and `district` are the codes of two location types on that domain).  When I do that, I get the following error message on save:

> Case List has invalid location reference state:county/name. Details: Reference type county cannot be a child of primary type state.

When I flip the order to `location:district:state/name`, it saves successfully.  If you try it with names that aren't real location types, it will give a different error.

Anyways, I think all we need to do for this migration is turn off that validation if the domain uses the flat fixture.  If I type gibberish in there it pretty much just lets it slide.

**Comment by czue** [comment]
thanks @esoergel. sorry i should have mentioned that @mkangia and i offlined about this.

> Anyways, I think all we need to do for this migration is turn off that validation if the domain uses the flat fixture.

I think we should actually just disallow it to use the "location:" syntax at all.

File: corehq/apps/app_manager/models.py
Comment (mkangia) [inline corehq/apps/app_manager/models.py]: @czue This would take care of discontinuing the feature to use location references in Case List. Any suggestions for better error message? Any alternative that could be suggested?

Comment (czue) [inline corehq/apps/app_manager/models.py]: Hmm. Maybe "That format is no longer supported. To reference the location hierarchy you need to use the "Custom Calculations in Case List	" feature preview. For more information see: https://confluence.dimagi.com/pages/viewpage.action?pageId=38276915". cc @sheelio

Comment (sheelio) [inline corehq/apps/app_manager/models.py]: I think that message+link is fine.

**Comment by esoergel** [comment]
Build failure was unrelated, just kicked it.

File: corehq/apps/locations/tests/test_location_fixtures.py
Comment (dimagimon) [inline corehq/apps/locations/tests/test_location_fixtures.py]: E303 too many blank lines (2)

**Review by esoergel — APPROVED** [review]

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (esoergel) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: Don't these domains also need the `HIERARCHICAL_LOCATION_FIXTURE` toggle enabled?  I agree that it makes sense to look at them individually, but I imagine in every case we'll want to be able to access the FixtureConfiguration UI and let them keep getting the hierarchical fixture.

Comment (mkangia) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: @esoergel @czue I would expect no domains to be in this area. These would be domains with flat fixture enabled but with configuration set to use hierarchical which contradicts. So this is just a sanity check.
We can always add them to the flag if needed but like i said there should not be any ideally.

File: corehq/apps/locations/management/commands/migrate_new_location_fixture.py
Comment (esoergel) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: Won't the `location_configuration` always be missing for these domains?  They're 
```python
domains_to_stay_on_hierarchical_fixture = (set(domains_having_locations) -
                                           set(domain_with_flat_fixture_enabled))
```
So if the flat fixture toggle isn't enabled, they shouldn't have a fixture config, right?

Comment (mkangia) [inline corehq/apps/locations/management/commands/migrate_new_location_fixture.py]: Yup! Thats right. Just logging for confirmation.

**Comment by mkangia** [comment]
@esoergel lets do this today!

**Pull request merged**
